### ISSUE TITLE
Investigate astrometry failures for Run1.2p

### ISSUE THREAD

**Issue by rearmstr** [LSSTDESC/ImageProcessingPipelines#71]
There are two sensors that had large astrometric residuals in 1.2p: 
263504-i 
`checkCcdAstrometry INFO: Processing {'visit': 263504, 'filter': 'i', 'raftName': 'R33', 'detectorName': 'S01', 'detector': 145}
checkCcdAstrometry INFO: Median astrometric scatter 207.06 mas`

227032-z
`checkCcdAstrometry INFO: Processing {'visit': 227032, 'filter': 'z', 'raftName': 'R32', 'detectorName': 'S20', 'detector': 141}
checkCcdAstrometry INFO: Median astrometric scatter 269.98 mas`

**Comment by johannct** [comment]
for  info: https://confluence.slac.stanford.edu/pages/viewpage.action?spaceKey=LSSTDESC&title=DC2+Run1.2+Astrometry+Failures

**Comment by rearmstr** [comment]
I was not able to reproduce the failure for 263504.  I was using software version w.2018.30.  If you also look at the logs on the failure page, it doesn't show a failure.  It would be helpful to have another example.  But I was able to find some things for the 227032 visit.

There are known issues with the astrometry being somewhat fragile.  In looking into this specific issues, the config parameters seem to be using the default values of the algorithm, not those from HSC.  I tried to use the HSC config values and they did not have the same problem.  The key difference in the config seems to be:

`config.calibrate.astrometry.matcher.sourceSelector.active.sourceFluxType = 'Psf'`

Using this resulted in finding twice the number of matched reference sources.  The default value of the fluxtype is the aperture flux.  So depending on the seeing and the size of the aperture this may not be the best thing to use.

I also tried to use the MatchPessimisticB algorithm and that also did not any problems in this case.  So we could consider using this instead of the default.

However it would be good to have some failures to see how robust these solutions are.   @johannct, what configurations did you use that resulted in the failures from 263504?

**Comment by johannct** [comment]
Best would be to check with @heather, who could reproduce the 2 failures at NERSC. I am recalling her post on confluence here : https://confluence.slac.stanford.edu/pages/viewpage.action?spaceKey=LSSTDESC&title=DC2+Run1.2+Astrometry+Failures

**Comment by johannct** [comment]
@rearmstr remember that one oddity is the fact that the failures did not occur for an imsim version of these 2 visits.
As for my config : w30 and obs_lsstCam dc2/run1.2i in Lyon : /sps//lsst/dataproducts/desc/DC2/Run1.2p/w_2018_30/rerun/210918/

**Comment by johannct** [comment]
>  Given  the fact that the astrometry failure rate is low, my recommendation  would be to handle the failures manually (by rerunning processCcd with MatchOptimisticB on the failed CCDs for instance).

not exactly a pleasant pipeline option.....

**Comment by boutigny** [comment]
> not exactly a pleasant pipeline option.....

Does a manual rerun break the pipeline ?
The other possibility is to modify `processCcd` to include `checkCcdAstrometry` and to rerun automatically the astrometry using a different matcher in case of failure. But I guess that this option would require a little bit of time to be implemented.

**Comment by johannct** [comment]
yes if I proceed with coadd automatically.... No if the operator remains a man-in-the-middle stage who needs to check output and act manually upon it at a certain step in the pipeline..... which is not a pipeline anymore in that case :)

**Comment by johannct** [comment]
I can relaunch processCcd in the pipeline based on checkCcdAstrometry's result. It adds complexity but it is doable if it is easy to parse checkCcdAstrometry output

**Comment by johannct** [comment]
but it has to be worth it....

**Comment by rearmstr** [comment]
Thanks for your comments.  Unless I am missing something obvious, I don't see the configuration changes you mention above in the current obs_lsstCam or ImageProcessPipelines code.  I think we should change the `sourceFluxType` to `Psf`  it seems to improve the number of matches in general and I think it makes sense to use that instead of a fixed aperture.  I wouldn't bother with the other config changes because as you say they are probably HSC specific.  I agree that this is not going to solve all the astrometry problems, the algorithm is known to be fragile and unreliable in certain cases, but if runs better using the psf flux we should certainly make the change.

I'm not totally surprised that there is no problem for imsim.  The observing conditions are different between the two pipelines and that can be enough for the matcher to have problems.  It is interesting that the failed CCD actually succeeds on the first two iterations.  It isn't until the third iteration that the matcher gets the wrong answer.

@boutigny, the default matcher is to be `MatchOptimisticB`, not `MatchPessimisticB`.  I'm guessing that is what you meant?

As @johannct says, we need to avoid doing any manual checks for DC2.  Instead of relaunching processCcd I think we could add an additional step to `checkCcdAstrometry` itself that would directly rerun the astrometry with `MatchPessimisticB` if needed.  We don't need to run processCcd again since the astrometry happens at the very end.

**Comment by rearmstr** [comment]
Not, offhand.  We could just pull out the corresponding bits from the `CalibrateTask`.

**Comment by boutigny** [comment]
@rearmstr I was probably unclear. The block of config parameters I was mentioning is in `obs_subaru/hsc` and is not in `obs_lsstCam`. I am surprised that it has not been copied over. So either it is an oversight or this block has been added recently to `obs_subaru/hsc`. And I agree that we should change  `sourceFluxType` to `Psf` if it improves the number of matches (even if I don't understand why it has a significant effect).
And you are right, I have mixed `MatchOptimisticB` and `MatchPessimisticB`
